src/pykeen/datasets/ea/combination.py

```diff
@@ -1,26 +1,174 @@
 """Combination strategies for entity alignment datasets."""
+
 import logging
-from abc import abstractmethod
-from typing import Dict, Tuple
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Iterable, Mapping, NamedTuple, Optional, Sequence, Tuple, Union
 
+import numpy
 import pandas
 import torch
 from class_resolver import ClassResolver
+from pandas.api.types import is_numeric_dtype, is_string_dtype
 
-from ...triples import TriplesFactory
-from ...typing import EA_SIDE_LEFT, EA_SIDE_RIGHT
-from ...utils import format_relative_comparison
+from ...triples import CoreTriplesFactory, TriplesFactory
+from ...typing import COLUMN_HEAD, COLUMN_TAIL, EA_SIDE_LEFT, EA_SIDE_RIGHT, EA_SIDES, MappedTriples, TargetColumn
+from ...utils import format_relative_comparison, get_connected_components
 
 logger = logging.getLogger(__name__)
 
 
-# TODO: support ID-only graphs
+def cat_triples(*triples: Union[CoreTriplesFactory, MappedTriples]) -> Tuple[MappedTriples, torch.LongTensor]:
+    """
+    Concatenate (shifted) triples.
+
+    :param triples:
+        the triples factories, or mapped triples
+
+    :return:
+        the concatenation of the shifted mapped triples such that there is no overlap, and the offsets for the
+        individual factories
+    """
+    # a buffer for the triples
+    res = []
+    # the overall offsets
+    offsets = torch.empty(len(triples), 3, dtype=torch.long)
+    # the current offset
+    offset = torch.zeros(1, 3, dtype=torch.long)
+    for i, x in enumerate(triples):
+        # store offset
+        offsets[i] = offset
+        # normalization
+        if isinstance(x, CoreTriplesFactory):
+            e_offset = x.num_entities
+            r_offset = x.num_relations
+            x = x.mapped_triples
+        else:
+            e_offset = x[:, [0, 2]].max().item() + 1
+            r_offset = x[:, 1].max().item() + 1
+        # append shifted mapped triples
+        res.append(x + offset)
+        # update offset
+        offset[:, 0::2] += e_offset
+        offset[:, 1] += r_offset
+    return torch.cat(res), offsets
+
+
+def _merge_mappings(
+    *pairs: Tuple[str, Mapping[str, int]],
+    offsets: torch.LongTensor = None,
+    mappings: Sequence[Mapping[int, int]] = None,
+    extra: Optional[Mapping[str, int]] = None,
+) -> Dict[str, int]:
+    result: Dict[str, int] = {}
+    for i, (prefix, mapping) in enumerate(pairs):
+        for key, value in mapping.items():
+            key = f"{prefix}:{key}"
+            if offsets:
```

Comment:
This throws `RuntimeError: Boolean value of Tensor with more than one value is ambiguous` and should probably be exchanged for `if offsets is not None`?

Reply:
5704533

Reply:
> After this change everything runs through smoothly and only the implementation of the smoketest is missing.

Did you already prepare that when you though about https://github.com/pykeen/pykeen/pull/893#discussion_r861553903 ? 😇

Reply:
Ah okay, that's what you meant, when you said you were waiting :sweat_smile:  No I didn't...